### ISSUE TITLE
ICML 26 | KVEraser: Learning to Steer KV Cache for Efficient Localized Context Erasing

### ISSUE THREAD

**Issue by Dszdsxc0405** [AI-Infra-Team/awesome-papers#214]
## 0. Information
Title：KVEraser: Learning to Steer KV Cache for Efficient Localized Context Erasing
Authors：Mufei Li, Shikun Liu, Dongqi Fu, Haoyu Wang, Yinglong Xia, Hong Li, Hong Yan, Pan Li
Institude：Georgia Institute of Technology；Meta
Link：https://arxiv.org/abs/2606.17034

## 1. Background and Problem

With the development of long-context large language models, RAG systems, agents, and tool-augmented applications, model inference increasingly relies on **KV cache** to reduce latency. KV cache stores the key/value states of each token after the model processes a context, so later generation does not need to recompute the entire context.

However, real-world long contexts are not always stable or reliable. For example, a RAG system may first retrieve and cache an outdated fact; a tool call may return an incorrect result; a user may retract a preference; or a harmful prompt injection may be mixed into the long context. These problematic contents are often discovered only after the model has already completed prefill and generated the KV cache.

This paper studies **post-hoc context erasing**: after a context has already been processed and cached, how can we efficiently remove a specific span so that the model behaves as if it had never seen that span?

The exact solution is **full recomputation**. Suppose the original context is:

```text
x = p + e + s
```

where `p` is the preserved prefix, `e` is the span to erase, and `s` is the suffix. After deletion, the ideal context becomes:

```text
x' = p + s
```

To obtain the fully correct KV cache, the system can reuse the KV cache of `p`, but it must recompute the entire suffix `s`. The problem is that `s` is often very long, while the span `e` to be erased may be very short. Therefore, the cost of exact erasing depends mainly on the suffix length rather than the erased-span length. This causes high latency in 16K, 32K, or even longer context scenarios.

<img width="1518" height="824" alt="Image" src="https://github.com/user-attachments/assets/309c978b-1e73-4b22-b6d4-ad6f0a9c4b63" />

---

## 2. Observations and Opportunity

The paper first points out that existing KV cache reuse and KV cache eviction methods cannot directly solve this problem.

**KV cache reuse** assumes that the reused cache remains a valid part of the target prompt. For example, if multiple requests share the same prefix, the system can directly reuse that prefix’s KV cache. However, in context erasing, the situation is the opposite: the span `e` to be removed is no longer valid content, so it cannot simply be reused.

**KV cache eviction** assumes that the removed KV states are usually unimportant or redundant. Its goal is to compress the cache and save memory while preserving generation quality. But in this paper’s setting, the span to erase may be invalid yet highly influential. For example, an incorrect factual passage may be the key evidence that causes the model to answer incorrectly. Therefore, it cannot be treated as a low-value token.

More importantly, the paper emphasizes that the influence of the erased span is not limited to its own KV cache. Because Transformer uses causal attention, every token in the suffix `s` was computed after seeing the preceding span `e`. As a result, even if we delete the KV cache corresponding to `e`, the KV cache of the suffix `s` may still carry the influence of `e`. The paper summarizes this as: **a local edit has a global effect**.

This is the core observation of the paper: deleting a short span looks like a local operation, but it contaminates the entire suffix cache. To completely remove this contamination, full recomputation must recompute the whole suffix. However, if we can find a way to modify the KV cache locally, we may significantly reduce inference cost.

The paper further explains this opportunity from a complexity perspective. The post-hoc cost of full recomputation is approximately:

```text
O(|s|(|p| + |s|))
```

If we only process the erased span `e`, the cost can become:

```text
O(|e|(|p| + |e|))
```

In practical applications, the erased span is usually much shorter than the suffix, i.e., `|e| << |s|`. For example, we may remove a 100-token erroneous retrieved passage, while the following document or conversation history may contain thousands or tens of thousands of tokens. Therefore, if local KV editing can approximate the behavior of full recomputation, it can bring substantial efficiency gains.

---

## 3. Innovation and Method

The paper proposes **KVEraser**, a learning-based KV cache editing algorithm. Its goal is not to reconstruct the complete KV cache after deletion, but to learn a local replacement block that makes the model behave similarly to the truly recomputed clean context.

The basic idea of KVEraser is to preserve the original KV cache of the prefix `p` and suffix `s`, and only replace the KV cache corresponding to the erased span `e`.

The original KV cache can be written as:

```text
KV(x) = KV(p) + KV(e) + KV(s)
```

KVEraser constructs a surrogate cache:

```text
KV_hat = KV(p) + KV_steer + KV(s)
```

Here, **KV_steer** is a set of steering states generated by a trainable eraser module. It is neither the original KV of `e` nor a simple blank placeholder. Instead, it is a learned key/value block. Its role is like a “corrective controller”: although the suffix KV may still contain contamination, the model’s later attention and generation are steered toward behavior close to that of the clean context after deletion.

<img width="1499" height="790" alt="Image" src="https://github.com/user-attachments/assets/b1016d9e-e559-4bc9-a6dc-4f3151cba7c5" />

KVEraser contains two main components:

1. **Frozen generator**: the original large language model. The paper uses Qwen3-8B. This model is frozen when training the eraser.
2. **Trainable eraser**: a trainable module `Eφ`, which generates the replacement steering KV based on the preserved prefix KV and the erased span.

The main inputs to the eraser are:

```text
Prefix KV + erased span e
```

The output is:

```text
KV_steer
```

The paper also emphasizes that KVEraser does not rely on the query when constructing the steering block. This makes the erasing operation **query-agnostic**. Once a span is erased, different future queries can reuse the same edited cache, without reconstructing a new steering block for every query.

For training, the paper adopts a two-stage strategy.

The first stage is **span-neighbor pre-training**. The authors construct 80K pre-training samples from Wikipedia. In each sample, they randomly insert a 100-token text chunk as the span to erase, and then ask the model to retrieve neighboring text correctly after that span is removed. The purpose of this stage is not to reconstruct the erased content, but to teach the eraser how to suppress the influence of the erased span on future behavior.

The second stage is **task-specific fine-tuning**. The authors fine-tune the eraser on two downstream scenarios: a controlled NIAH synthetic task and a long-document QA task with misleading factual distractors. The QA training samples come from Natural Questions, TriviaQA, and HotpotQA. The authors only keep samples where the inserted distractor actually causes the model to answer incorrectly, ensuring that the erased content is genuinely harmful rather than merely irrelevant.

---

## 4. Experiments and Results

The paper evaluates KVEraser in two main settings: a controlled needle-in-a-haystack erasing task, and factual distractor erasing in real long-document QA.

In the NIAH experiment, each sample contains two “magic number” sentences with the same key but different values. The earlier needle is designated as the span to erase, while the later needle is retained. After erasing, the model should answer with the retained value. The advantage of this task is that the values are randomly generated, so the model cannot rely on parametric knowledge; it must rely on the context and the behavior of the erased KV cache.

The experiment covers context lengths of 1K, 2K, 4K, 8K, 16K, and 32K. The results show that KVEraser achieves nearly perfect exact match across all context lengths, almost matching full recomputation. Other approximate methods perform much worse: delete-and-shift almost completely fails; instruction-only forgetting works somewhat in short contexts but degrades quickly as the context becomes longer; local suffix repair also fails to recover correct behavior reliably.

In terms of efficiency, KVEraser shows a clear advantage. As the context length increases from 1K to 32K, the latency of KVEraser increases by only 24%, while the latency of full recomputation increases by 17.6x. This confirms that the computation of KVEraser mainly depends on the erased span rather than the entire suffix.

<img width="1479" height="772" alt="Image" src="https://github.com/user-attachments/assets/8ad42693-653b-4f17-8344-c6fac5bfb136" />

In the real QA experiment, the authors use three datasets unseen during training: 2WikiMultiHopQA, MuSiQue, and IIRC. In each sample, a misleading factual distractor is inserted, and the method is expected to erase this distractor before answering the question.

The results show that full recomputation still achieves the highest exact match, but its latency is 3-4x that of KVEraser. Among all approximate methods, KVEraser achieves the highest exact match while maintaining lower or comparable latency. Therefore, it lies on the quality-efficiency Pareto frontier. In other words, if we do not accept the high latency of full recomputation, KVEraser provides the best trade-off.

<img width="1464" height="717" alt="Image" src="https://github.com/user-attachments/assets/fc9a1b08-af81-4e26-980f-55c615df0b88" />

The paper also conducts an ablation study on the information sources used by the eraser. The full KVEraser uses both prefix KV and the erased span, achieving an average exact match of 0.879. Removing the prefix reduces the score to 0.859. Adding the query gives an average score of 0.871, and adding 15% suffix KV gives 0.844. These results show that prefix KV plus the erased span already contains the most important information, while adding query or suffix information does not bring consistent gains.

<img width="1482" height="283" alt="Image" src="https://github.com/user-attachments/assets/e9f47bd2-8f26-4a2d-8d56-eacdee677d46" />

Overall, the main contribution of KVEraser is that it systematically studies the problem of post-hoc context erasing in long-context LLMs and proposes a learning-based local KV editing method. It avoids recomputing the long suffix while still approximating the behavior of full recomputation across multiple tasks. This method has potential applications in RAG systems, long-running agents, tool-augmented models, and safe context management. The paper also notes that KVEraser still has a gap compared with full recomputation in natural QA, and future work may further improve it by scaling up training data and increasing data diversity.
